### PR TITLE
Fixes #1454. Now MAPL_Shmem memory is deallocated properly

### LINKED ISSUE #1454: GenericCplComp.F90 doest not deallocate tilemask properly
@sdrabenh reported a crash when outputting monthly averages from 'tile' collection.

29593  NOT using buffer I/O for file: geosgcm_landice_checkpoint
29594 forrtl: severe (173): A pointer passed to DEALLOCATE points to an object that cannot be deallocated
29595 Image              PC                Routine            Line        Source·············
29596 GEOSgcm.x          00000000034090D1  Unknown               Unknown  Unknown
29597 libMAPL.generic.s  00002AAAB633A755  mapl_genericcplco        1535  GenericCplComp.F90
29598 libesmf.so         00002AAABB6969E4  _ZN5ESMCI6FTable1     Unknown  Unknown

The problem is the tilemask is allocated conditionally using MAPL_Shmem or on the heap, but not deallocated accordingly

### PR DESCRIPTION
**Pull request by atrayano**
## Types of changes

- [x ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [x ] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x ] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x ] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

File: generic/GenericCplComp.F90 [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#define DEALOC_(A) if(associated(A))then;if(MAPL_ShmInitialized)then;call MAPL_SyncSharedMemory(rc=STATUS);call MAPL_DeAllocNodeArray(A,rc=STATUS);else;deallocate(A,stat=STATUS);endif;_VERIFY(STATUS);NULLIFY(A);endif
```
Comment (tclune) [inline generic/GenericCplComp.F90:1]: If we are going to make hideous macros, let's at least make them readable.  You'll need to check that I did not break anything in the edits though.
```suggestion
#define DEALOC_(A) \
    if(associated(A))then \
          if(MAPL_ShmInitialized)then \
              call MAPL_SyncSharedMemory(rc=STATUS) \
              call MAPL_DeAllocNodeArray(A,rc=STATUS) \
           else \
              deallocate(A,stat=STATUS) \
           endif \
       _VERIFY(STATUS) \
       NULLIFY(A) \
    endif
```

File: generic/GenericCplComp.F90 [hunk 2]
```diff
@@ -1338,7 +1340,7 @@ subroutine ReadRestart(CC, SRC, DST, CLOCK, RC)
           case default
              _ASSERT(.false., "Unsupported rank")
           end select
-          if(associated(mask)) deallocate(mask)
+          DEALOC_(mask)
```
Comment (tclune) [inline generic/GenericCplComp.F90:1343]: I also suggest renaming the macro to have a leading underscore `_DEALOC` unless that is too different than existing practice.

File: generic/GenericCplComp.F90 (hunk 1 above)
Comment (atrayano) [reply generic/GenericCplComp.F90:1]: This is much cleaner! I wish we have done very similar macros the same way.

Comment (tclune) [reply generic/GenericCplComp.F90:1]: I only recently became aware that this was portable.

Comment (atrayano) [reply generic/GenericCplComp.F90:1]: The intel compiler did not like the new style. Let me try with ; at the end of each line

Comment (tclune) [reply generic/GenericCplComp.F90:1]: Sorry about that.

Comment (atrayano) [reply generic/GenericCplComp.F90:1]: ";" works. I am adopting the new style, and about to run a quick test, and push the changes

File: generic/GenericCplComp.F90 (hunk 2 above)
Comment (atrayano) [reply generic/GenericCplComp.F90:1343]: _DEALLOC works

**Review by tclune — DISMISSED** [review]

**Comment by mathomp4** [comment]
Let me do a sanity check before I issue the release. For now, I'm going to hijack this PR to prepare for a release. @tclune might need to re-approve.

**Review by mathomp4 — APPROVED** [review]
Looks good!

**Pull request merged**